tests/unit/test_p11wrapper.cpp

```diff
@@ -0,0 +1,403 @@
+};
+
+namespace testutils
+{
+PKCS11_CTX *somePKCS11CtxPtr()
+{
+    /* Reserve some memory and cast a pointer to that ; pointers will not be dereferenced */
+    static char dummyBuf[42] = {};
+    return reinterpret_cast<PKCS11_CTX *>(&dummyBuf);
+}
+
+PKCS11_SLOT *somePKCS11SlotPtr()
+{
+    /* Reserve some memory and cast a pointer to that ; pointers will not be dereferenced */
+    static char dummyBuf[42] = {};
+    return reinterpret_cast<PKCS11_SLOT *>(&dummyBuf);
+}
+
+PKCS11_TOKEN *somePKCS11TokenPtr()
+{
+    /* Reserve some memory and cast a pointer to that ; pointers will not be dereferenced */
+    static char dummyBuf[42] = {};
+    return reinterpret_cast<PKCS11_TOKEN *>(&dummyBuf);
+}
+
+EVP_PKEY *someEVP_PKEY()
+{
+    /* Reserve some memory and cast a pointer to that ; pointers will not be dereferenced */
+    static char dummyBuf[42] = {};
+    return reinterpret_cast<EVP_PKEY *>(&dummyBuf);
+}
+
+}  // namespace testutils
+
+void P11WrapperTest::SetUp() { LibP11MockManager::resetMock(); }
+
+void P11WrapperTest::TearDown() { LibP11MockManager::destroy(); }
+
+/* Test that exception is triggered if P11_PKCS11_CTX_new()
+ * returns NULL.
+ */
+TEST_F(P11WrapperTest, badInitPKCS11Ctx)
+{
+    EXPECT_CALL(_mock(), P11_PKCS11_CTX_new()).WillOnce(Return(nullptr));
+
+    EXPECT_THROW(_PKCS11_CTX_new(), P11Exception);
+}
+
+/* Test successful _PKCS11_CTX_new(). */
+TEST_F(P11WrapperTest, goodInitPKCS11Ctx)
+{
+    EXPECT_CALL(_mock(), P11_PKCS11_CTX_new()).WillOnce(Return(::testutils::somePKCS11CtxPtr()));
+
+    /* Since we wrap in a unique_ptr, expect a call to the "free" function */
+    EXPECT_CALL(_mock(), P11_PKCS11_CTX_free(::testutils::somePKCS11CtxPtr()));
+
+    EXPECT_NO_THROW(_PKCS11_CTX_new());
+}
+
+/* Test failed _PKCS11_CTX_load(). */
+TEST_F(P11WrapperTest, badPKCS11Load)
+{
+    std::string test = "test";
+
+    EXPECT_CALL(_mock(), P11_PKCS11_CTX_load(nullptr, test.c_str())).WillOnce(Return(-1));
+
+    EXPECT_THROW(_PKCS11_CTX_load(nullptr, test), P11Exception);
+}
+
+/* Test successful _PKCS11_CTX_load(). */
+TEST_F(P11WrapperTest, goodPKCS11Load)
+{
+    std::string test = "test";
+    EXPECT_CALL(_mock(), P11_PKCS11_CTX_load(::testutils::somePKCS11CtxPtr(), test.c_str()))
+            .WillOnce(Return(0));
+
+    EXPECT_NO_THROW(_PKCS11_CTX_load(::testutils::somePKCS11CtxPtr(), test));
+}
+
+/* Test  _PKCS11_enumerate_slots() with NULL context. */
+TEST_F(P11WrapperTest, badSlotEnumeration)
+{
+    EXPECT_THROW(_PKCS11_enumerate_slots(nullptr), P11Exception);
+}
+
+/* Test _PKCS11_enumerate_slots with valid context. */
+TEST_F(P11WrapperTest, goodSlotEnumeration)
+{
+    EXPECT_CALL(_mock(), P11_PKCS11_enumerate_slots(::testutils::somePKCS11CtxPtr(), _, _))
+            .WillOnce(Return(0));
+
+    EXPECT_NO_THROW(_PKCS11_enumerate_slots(::testutils::somePKCS11CtxPtr()));
+}
+
+/* Test _PKCS11_release_all_slots() with NULL context. */
```

Comment:
Code duplication.
Can you please provide the test parameters to a single function?

Reply:
Done.